### ISSUE TITLE
compile-time use of undefined value detection

### ISSUE THREAD

**Issue by andrewrk** [zig-lang/zig#597]
```zig
extern fn bar();
export fn entry() {
    var a: i32 = undefined;
    if (a > 10) {
        bar();
    }
}
```

Zig should be able to detect that we will with 100% certainty branch on an undefined value.

Related:

```zig
export fn entry() {
    unreachable;
}
```

Should be an error, because it's 100% certain at compile-time that we will reach unreachable code.

Not to be confused with #211 which is adding extra code and data in safety-enabled builds to detect use of undefined value at runtime.

**Comment by pluto439** [comment]
After `a = undefined`, will I be able to create another variable with the same name, but a different type?

**Comment by andrewrk** [comment]
No, you would need to make `a` go out of scope. This use case is not supported because we want code readers to have the guarantee that within a scope, if you see a variable name, that same variable name is the same type and refers to the same memory location in the entire scope.

**Comment by pluto439** [comment]
It's not the same as deleting a variable then. I want it to stop existing, completely. I will only ever delete variables that are temporary anyway.

**Comment by PavelVozenilek** [comment]
@pluto439:

> After a = undefined, will I be able to create another variable with the same name, but a different type?

You are asking for feature similar to "hyperstatic scope" used in Forths ( http://wiki.c2.com/?HyperStaticGlobalEnvironment ). Here it is used for functions (Forth doesn't have anything else) and allows to reuse their names, reducing the need to invent unique words all the time. It is one of Forth's strong points.

**Comment by pluto439** [comment]
@PavelVozenilek

That's too smart for me. I'm just trying to steal the `del` keyword from python. Which in c-like languages will mean "free stack, or mark this place on stack as free to reuse".

**Comment by PavelVozenilek** [comment]
@pluto439

> Which in c-like languages will mean "free stack, or mark this place on stack as free to reuse".

LLVM does this automatically, it is called "stack slot coloring".

**Comment by pluto439** [comment]
@PavelVozenilek I know, but didn't knew it's called "stack slot coloring". Just wanted to say that it's possible for compiled languages too, and not for interpreted ones only.

It's more of a hint for a programmer that this variable will not be used again, and also a safety net. More details here #594, further discussion should be there I guess.

**Comment by andrewrk** [comment]
Related: #1947
